core/templates/dev/head/player/InteractionService.js

```diff
@@ -0,0 +1,94 @@
+// Copyright 2015 The Oppia Authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS-IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/**
+ * @fileoverview Shared utilities for interactions.
+ *
+ * @author wxyxinyu@gmail.com (Xinyu Wu)
+ */
+
+oppia.constant('DEFAULT_ANSWER_GROUP_CLASSIFICATION_THRESHOLD', 0.3);
+
+oppia.factory('interactionService', [
+    'warningsData', 'DEFAULT_ANSWER_GROUP_CLASSIFICATION_THRESHOLD', function(
+    warningsData, DEFAULT_ANSWER_GROUP_CLASSIFICATION_THRESHOLD) {
+  var stringifyClassifiedRule = function(rule) {
+    var paramList = [];
+    for (var key in rule.inputs) {
```

Comment:
IIRC, the order will only damage the strings used in the answer submission event handler, so it might mess up some statistics. Either way, I don't like stringifying RuleSpecs and this might be deprecated down the road (after the new answers structure is in place).

That being said, this is duplicated on the backend. Although a lot of code will be duplicated as the answers framework is moved to the frontend,  I think this particular code should remain untouched and stay in the backend. You will need to update the reader.AnswerSubmittedEventHandler to accept an answer group index and rule spec index so that you may retrieve the rulespec and stringify in that handler. I think it will be nicer to detach the concept of 'rule spec string' from the frontend entirely. This will make it easier to remove the concept entirely from the codebase later on.